introduce filters and tags for server policy load balancing

martinfurmanski opened the pull request:
This is the first step in building up the server-side policy based
load balancing strategy. The idea is that servers are tagged
with one or more tags and these are used in a server-side configured
scheme of filters to select those which match. Different policies can
select different sets of tags and there can be other filters attached apart
from tag-based filters.

Every policy has a name which the client can use to bind to a particular
policy by supplying that name in the client supplied context, as supported
by the V2 version of GetServers.

martinfurmanski committed:
introduce filters and tags for server policy load balancing

This is the first step in building up the server-side policy based
load balancing strategy. The idea is that servers are tagged
with one or more tags and these are used in a server-side configured
scheme of filters to select those which match. Different policies can
select different sets of tags and there can be other filters attached
apart from tag-based filters.

Every policy has a name which the client can use to bind to a particular
policy by supplying that name in the client supplied context, as supported
by the V2 version of GetServers.

jimwebber commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/discovery/ClusterTopology.java:
```diff
@@ -36,8 +36,8 @@ public ClusterTopology( CoreTopology coreTopology, ReadReplicaTopology readRepli
     {
-        Optional<CoreAddresses> coreAddresses = coreTopology.find( upstream );
-        Optional<ReadReplicaAddresses> readReplicaAddresses = readReplicaTopology.findAddressFor( upstream );
+        Optional<CoreServerInfo> coreAddresses = coreTopology.find( upstream );
```
rename the local variables too?

jimwebber commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/discovery/HazelcastClient.java:
```diff
@@ -37,6 +39,9 @@
 import static org.neo4j.causalclustering.discovery.HazelcastClusterTopology.READ_REPLICA_BOLT_ADDRESS_MAP_NAME;
 import static org.neo4j.causalclustering.discovery.HazelcastClusterTopology.READ_REPLICA_MEMBER_ID_MAP_NAME;
 import static org.neo4j.causalclustering.discovery.HazelcastClusterTopology.READ_REPLICA_TRANSACTION_SERVER_ADDRESS_MAP_NAME;
+import static org.neo4j.causalclustering.discovery.HazelcastClusterTopology.SERVER_TAGS_MULTIMAP_NAME;
```
Doesn't this leak information from the plugin into the HZ instance which should be unaware of plugins?

jimwebber replied:
Should the plugins get a  `ReadReplicaTopologyService`instance injected instead?

martinfurmanski replied:
I think tagging could be seen as a general concept for identifying servers or sets of servers, and you could imagine building several things on top of that, e.g. load balancing and replication policies, some future security policy, general affinity, etc. So from that perspective I don't think the layering is bad.

If you look at the setting I simply named it `causal_clustering.server_tags` which is inline with this way of looking at it (e.g. no load_balancing in the name).

Is tagging really the most appropriate design for identification meant for other purposes? Future will have to show.

martinfurmanski replied:
On the second point, yes, and I will rename that to TopologyService.

martinfurmanski committed:
cleanup and clarify around topology/discovery

The topology service is an internal service provided by the discovery subsystem,
also referred to as the discovery service.

The TopologyService provides a read-only interface for reading the eventually
consistent topology-related information, while the CoreTopologyService extends
upon that with additional methods required in special circumstances which
require support from the underlying discovery service.

The distinction was not clear from the naming nor documented. This commit
is a first step towards better clarity.

ReadReplicaTopologyService was renamed simply to TopologyService since the
distinction is not on who the client of the service is, but rather on
the width of the interface required. There are clients of TopologyService
both in cores and read replicas, and it can give the complete topology, thus
naming it with ReadReplica makes little sense.
